tests/utils/test_session_token.py

```diff
@@ -0,0 +1,111 @@
+from datetime import datetime, timedelta
+
+import jwt
+import pytest
+
+from spylib.utils.session_token import (
+    InvalidIssuerError,
+    MismatchedHostError,
+    SessionToken,
+    TokenAuthenticationError,
+)
+
+api_key = "API_KEY"
+api_secret = "API_SECRET"
+now = datetime.now()
+
+
+def get_token():
+    """
+    This function returns the same data over and over again. We don't use a variable
+    as the fixtures will share the memory and overwrite the value, even with proper scopes.
+    """
+    return {
+        "iss": "https://test.myshopify.com/admin",
+        "dest": "https://test.myshopify.com",
+        "aud": api_key,
+        "sub": 1,
+        "exp": (now + timedelta(0, 60)).timestamp(),
+        "nbf": (now - timedelta(0, 60)).timestamp(),
+        "iat": now.timestamp(),
+        "jti": 123,
+        "sid": "abc123",
+    }
+
+
+@pytest.fixture(scope='function', autouse=False)
+def token():
+    return get_token()
+
+
+def generate_auth_header(token):
+    return f'Bearer {jwt.encode(token, api_secret, algorithm="HS256")}'
+
+
+@pytest.mark.asyncio
+async def test_session_token(token):
+    valid_auth_header = generate_auth_header(token)
+    session_token = SessionToken.decode_token_from_header(valid_auth_header, api_key, api_secret)
+
+    assert session_token.iss == token['iss']
+    assert session_token.dest == token['dest']
+    assert session_token.aud == token['aud']
+    assert session_token.sub == token['sub']
+    assert session_token.exp == token['exp']
+    assert session_token.nbf == token['nbf']
+    assert session_token.iat == token['iat']
+    assert session_token.jti == token['jti']
+    assert session_token.sid == token['sid']
+
+
+@pytest.mark.asyncio
+async def test_invalid_header():
+    with pytest.raises(TokenAuthenticationError):
+        SessionToken.decode_token_from_header('', api_key, api_secret)
+
+
+@pytest.mark.asyncio
+async def test_invalid_signature(token):
+    header = f'Bearer {jwt.encode(token, "invalid_secret", algorithm="HS256")}'
+    with pytest.raises(jwt.InvalidSignatureError):
+        SessionToken.decode_token_from_header(header, api_key, api_secret)
+
+
+@pytest.mark.asyncio
+async def test_invalid_hostname(token):
+    token['iss'] = 'https://someinvalidhost.com'
+    header = generate_auth_header(token)
+    with pytest.raises(InvalidIssuerError):
+        SessionToken.decode_token_from_header(header, api_key, api_secret)
```

Comment:
You should be able to parametrize this test such that you don't have to rewrite it 5 times ... copy-pasting is evil.
```suggestion
@pytest.mark.asyncio
async def test_invalid_hostname(token, fixture):
    token.update(fixture.token_mod)  # with token_mod = dict(iss='https://someinvalidhost.com')
    header = fixture.header  # My understanding is that token is always the same ... so this could be generated once as well, and could be '' if needed.
    with pytest.raises(fixture.expected_exception):
        SessionToken.decode_token_from_header(header, api_key, api_secret)
```